runtime/src/main/kotlin/kotlin/native/concurrent/WorkerBoundReference.kt

```diff
@@ -48,10 +48,16 @@ public class WorkerBoundReference<out T : Any>(value: T) {
      * The referenced value or null if referred object is not frozen and current worker is different from the one created [this].
      */
     val valueOrNull: T?
-        get() = @Suppress("UNCHECKED_CAST") (derefWorkerBoundReference(ptr) as T?)
+        get() = if (valueBeforeFreezing != null) valueBeforeFreezing else @Suppress("UNCHECKED_CAST") (derefWorkerBoundReference(ptr) as T?)
 
     /**
      * Worker that [value] is bound to.
      */
     val worker: Worker = Worker.current
+
+    @ExportForCppRuntime("Kotlin_WorkerBoundReference_doFreeze")
+    private fun doFreeze() {
+        ptr = createWorkerBoundReference(valueBeforeFreezing!!)
+        valueBeforeFreezing = null
```

Comment:
The implementation should probably be idempotent, since freezing can fail after the hooks are already applied. Then next attempt to freeze will run the hooks again.

Reply:
Yep, thanks for catching!